sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/transforms/Group.java

```diff
@@ -0,0 +1,684 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.sdk.schemas.transforms;
+
+import java.util.List;
+import javax.annotation.Nullable;
+import org.apache.beam.sdk.annotations.Experimental;
+import org.apache.beam.sdk.annotations.Experimental.Kind;
+import org.apache.beam.sdk.coders.KvCoder;
+import org.apache.beam.sdk.schemas.FieldAccessDescriptor;
+import org.apache.beam.sdk.schemas.Schema;
+import org.apache.beam.sdk.schemas.Schema.Field;
+import org.apache.beam.sdk.schemas.SchemaCoder;
+import org.apache.beam.sdk.transforms.Combine;
+import org.apache.beam.sdk.transforms.Combine.CombineFn;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.GroupByKey;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.Values;
+import org.apache.beam.sdk.transforms.WithKeys;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.Row;
+
+/**
+ * A generic grouping transform for schema {@link PCollection}s.
+ *
+ * <p>When used without a combiner, this transforms simply acts as a {@link GroupByKey} but without
+ * the need for the user to explicitly extract the keys. For example, consider the following input
+ * type:
+ *
+ * <pre>{@code
+ * {@literal @DefaultSchema(JavaFieldSchema.class)}
+ * public class UserPurchase {
+ *   public String userId;
+ *   public String country;
+ *   public long cost;
+ *   public double transactionDuration;
+ * }
+ *
+ * {@literal PCollection<UserPurchase>} purchases = readUserPurchases();
+ * }</pre>
+ *
+ * <p>You can group all purchases by user and country as follows:
+ *
+ * <pre>{@code
+ * {@literal @DefaultSchema(JavaFieldSchema.class)}
+ * {@literal PCollection<KV<Row, Iterable<UserPurchase>>} byUser =
+ *   purchases.apply(Group.byFieldNames("userId', "country"));
+ * }</pre>
+ *
+ * <p>However often an aggregation of some form is desired. The builder methods inside the Group
+ * class allows building up separate aggregations for every field (or set of fields) on the input
+ * schema, and generating an output schema based on these aggregations. For example:
+ *
+ * <pre>{@code
+ * {@literal PCollection<KV<Row, Row>>} aggregated = purchases
+ *      .apply(Group.byFieldNames("userId', "country")
+ *          .aggregateField("cost", Sum.ofLongs(), "total_cost")
```

Comment:
I think the syntax is a bit confusing. It mixes up SQL-like `GROUP BY` with Beam's `GroupByKey`, so for me it takes a second to figure out what this translates to. E.g. if I read this coming from SQL, it's unclear why I am grouping by `userId` and `country` but then aggregating only `cost`, but not grouping by `cost`.

It's also unclear what the output schema here is. Is it only the fields in the `Group.`? How does it work with `Select`?

Would it be better to follow SQL-like approach? I.e. add aggregations in `Select` and use `Group` to only specify the keys?

Reply:
This is closer to the Dataframes model, which I think is what we should be aiming at for a programmatic approach. It's also what fits in with Beam, where aggregations and groupbykey are intimately related. I don't think these transforms have to match SQL - we already have SQL for that scenario.

The output schema is specified by the aggregateField calls (that's why you have to specify a field name in those calls).

Reply:
I think that we definitely should have support for DataFrames model, but am not sure whether we should adjust internal Beam functionality to partially match it to get a similar feel in some places (but not the others). I'd rather keep everything Beam-specific here, and implement a more robust DataFrames model in a separate layer, and then maybe adjust Beam model if there are common things

Reply:
This is not a DataFrames API here. I think we should have one, but it might make more sense as a layer on top of Calcite.